src/main/java/duke/commands/DukeCommand.java

```diff
@@ -0,0 +1,70 @@
+package duke.commands;
+
+import duke.exceptions.DukeException;
+import duke.exceptions.DukeExceptionCommandNotFound;
+import duke.exceptions.DukeExceptionIllegalArgument;
+import duke.storage.FileLoader;
+import duke.tasks.Event;
+import duke.tasks.Todo;
+import duke.tasks.Deadline;
+import duke.tasks.Task;
+import duke.tasks.TaskList;
+import duke.ui.Ui;
+
+public abstract class DukeCommand {
+
+    public boolean isExit() {
+        return false;
+    }
+
+    /**
+     * Returns command based on user input line.
+     *
+     * The first-hop parsing occurs here.
+     * The rest of the parsing is offloaded to the individual parsing classes.
+     *
+     * @param line User input string.
+     * @return DukeCommand.
+     * @throws DukeExceptionCommandNotFound When command supplied is invalid.
+     * @throws DukeExceptionIllegalArgument When task parsing error occurs.
+     */
+    public static DukeCommand parse(String line)
+            throws DukeExceptionCommandNotFound, DukeExceptionIllegalArgument {
+        line = line.strip(); // input sanitization
+        String arg = "", cmd = line;
+        if (line.contains(" ")) {
+            int splitIdx = line.indexOf(' ');
+            cmd = line.substring(0, splitIdx); // automatically stripped
+            arg = line.substring(splitIdx+1).strip();
+        }
+        cmd = cmd.toLowerCase();
+
+        Task task;
+        switch (cmd) {
+        case "bye":
+            return new DukeCommandBye();
```

Comment:
Not a coding standard violation, but this is a cyclic dependency, perhaps there's another way to do this? (Not that this is bad but just a comment!)

Reply:
Good comment actually. This whole SRP and cyclic dependency is quite problematic for me.
I've recently updated the codebase to use Factory method for a similar issue with `Task <- Todo, Deadline, Event` and replacing a `Task.parseStringAsTask` with:

1. `TaskFileStringParser <- EventFileStringParser`: Offload parsing
2. `GetFileStringParser`: To select appropriate parser (dependency on tasks)

which spawned several more files, a whole lot of message passing, as well as additional javadoc requirements to fulfill. Not sure if this overhead is really worth the effort to decouple the classes.